fluss-connectors/fluss-connector-flink/src/main/java/com/alibaba/fluss/connector/flink/source/enumerator/initializer/BucketOffsetsRetrieverImpl.java

```diff
@@ -62,7 +62,26 @@ public Map<Integer, Long> earliestOffsets(
     @Override
     public Map<Integer, Long> offsetsFromTimestamp(
             @Nullable String partitionName, Collection<Integer> buckets, long timestamp) {
-        return listOffsets(partitionName, buckets, new OffsetSpec.TimestampSpec(timestamp));
+
+        // First get the current end offsets of the buckets. This is going to be used in case we
+        // cannot find a suitable offsets based on the timestamp, i.e. the message meeting the
+        // requirement of the timestamp have not been produced to fluss yet. In this case, we just
+        // use the latest offset. We need to get the latest offsets before querying offsets by time
+        // to ensure that no message is going to be missed.
+        Map<Integer, Long> endOffsets = latestOffsets(partitionName, buckets);
```

Comment:
I recommend this fallback logic should be handled in server-side to have a data consistency and simplify client. 

From the perspective of `ListOffset(timestamp)` RPC method semantics, it should "return the offset of the first message whose timestamp is greater than or equals to the given timestamp. **If no such message is found, the log end offset is returned.**" We can still throw `InvalidTimestampException` if the timestamp is larger than current timestamp (a future time). 

And then, adds unit tests in `ReplicaManagerTest#testListOffsetsWithTimestamp`.

This is also how kafka implement it: https://github.com/apache/kafka/blob/b40a962a8ead04b10f7a8eba346fe79784de9e4e/core/src/main/scala/kafka/log/UnifiedLog.scala#L1252-L1256